Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java

```diff
@@ -14,10 +14,15 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 import static com.earth2me.essentials.I18n.tl;
 
 public class Commandpay extends EssentialsLoopCommand {
+    private static final Pattern AMOUNT_FORMAT = Pattern.compile("((?:[0-9][_'`,]?)+(?:\\.[0-9]+)?|\\.[0-9]+)([kmbt])?");
```

Comment:
I have concerns that this regex isn't doing anything particularly useful here. It isn't strict enough to prevent non-numbers from being entered and introduces some assumptions as md mentioned about what a valid number separator is. Perhaps it would be "good enough" to say that there can't be any letters as part of the number except for the suffix? Then you remove the suffix part and sanitize the rest for the number part.

I'm also pretty sure the `\\.` here is probably intended to be `\.` since otherwise you're matching backslashes and not periods.

Reply:
Not sure what do you mean by "isn't strict enough to prevent non-numbers" tbh. The only concern I see here there is [this one](https://github.com/EssentialsX/Essentials/pull/5507#issuecomment-1716591852), but I don't think it's a major one.
As for allowing everything non-letter, I'm very 50/50 about it. E.g. if a server has a forum/discord/etc and somebody asks there to copy and use the command `/pay 1ɢᴇm`, it will be interpreted as `1m`/`1000000`, since `ɢᴇ` are simply unicode characters.

> I'm also pretty sure the `\\.` here is probably intended to be `\.` since otherwise you're matching backslashes and not periods.

The reason why there's two backslashes is that Java requires backslash to be escaped. To match backslash you'll actually need to type four of those - `\\\\`.

Reply:
I'm referencing mostly what you mentioned above (e.g. ``1,2_3`4'.5`` being valid, but also stuff like `1,0,0,0.23`). It feels like you would want to be stricter here if the goal is to block anything potentially ambiguous, but maybe I'm just nit picking too much. Probably unlikely to become an issue for anyone.